spv/sync.go

```diff
@@ -1155,110 +1290,143 @@ func (s *Syncer) getHeaders(ctx context.Context, rp *p2p.RemotePeer) error {
+	for walletID, w := range s.wallets {
+		s.fetchMissingCfiltersStart(walletID)
+		progress := make(chan wallet.MissingCFilterProgress, 1)
+		go w.FetchMissingCFiltersWithProgress(ctx, rp, progress)
+
+		for p := range progress {
+			if p.Err != nil {
+				return p.Err
+			}
+			s.fetchMissingCfiltersProgress(walletID, p.BlockHeightStart, p.BlockHeightEnd)
+		}
+		s.fetchMissingCfiltersFinished(walletID)
+	}
+	return nil
+}
+
 func (s *Syncer) startupSync(ctx context.Context, rp *p2p.RemotePeer) error {
+	log.Infof("Syncing %d wallets", len(s.wallets))
+
+	_, tipHeight, _ := s.highestChainTip(ctx)
+
 	// Disconnect from the peer if their advertised block height is
-	// significantly behind the wallet's.
-	_, tipHeight := s.wallet.MainChainTip(ctx)
+	// significantly behind the highest block height recorded by all wallets.
 	if rp.InitialHeight() < tipHeight-6 {
 		return errors.E("peer is not synced")
 	}
-	s.fetchMissingCfiltersStart()
-	progress := make(chan wallet.MissingCFilterProgress, 1)
-	go s.wallet.FetchMissingCFiltersWithProgress(ctx, rp, progress)
 
-	for p := range progress {
-		if p.Err != nil {
-			return p.Err
-		}
-		s.fetchMissingCfiltersProgress(p.BlockHeightStart, p.BlockHeightEnd)
+	if err := s.fetchMissingCFilters(ctx, rp); err != nil {
+		return err
 	}
-	s.fetchMissingCfiltersFinished()
 
 	// Fetch any unseen headers from the peer.
-	s.fetchHeadersStart()
+	s.fetchHeadersStart(rp.InitialHeight())
 	log.Debugf("Fetching headers from %v", rp.RemoteAddr())
 	err := s.getHeaders(ctx, rp)
 	if err != nil {
 		return err
 	}
 	s.fetchHeadersFinished()
+	log.Debugf("Finished fetching headers from %v", rp.RemoteAddr())
+
+	for {
+		walletID, w := s.nextWalletToRescan()
+		if w == nil {
+			break
+		}
 
-	if atomic.CompareAndSwapUint32(&s.atomicCatchUpTryLock, 0, 1) {
 		err = func() error {
-			rescanPoint, err := s.wallet.RescanPoint(ctx, )
+			rescanPoint, err := w.RescanPoint(ctx)
 			if err != nil {
 				return err
 			}
+			walletBackend := &WalletBackend{
+				Syncer:   s,
+				WalletID: walletID,
+			}
 			if rescanPoint == nil {
-				if !s.loadedFilters {
-					err = s.wallet.LoadActiveDataFilters(ctx, s, true)
+				if !s.loadedFilters[walletID] {
+					err = w.LoadActiveDataFilters(ctx, walletBackend, true)
 					if err != nil {
 						return err
 					}
-					s.loadedFilters = true
+					s.loadedFilters[walletID] = true
 				}
 
-				s.synced()
+				s.synced(walletID)
 
 				return nil
 			}
 			// RescanPoint is != nil so we are not synced to the peer and
 			// check to see if it was previously synced
-			s.unsynced()
+			s.unsynced(walletID)
 
-			s.discoverAddressesStart()
-			err = s.wallet.DiscoverActiveAddresses(ctx, rp, rescanPoint, s.discoverAccounts)
+			s.discoverAddressesStart(walletID)
+			err = w.DiscoverActiveAddresses(ctx, rp, rescanPoint, !w.Locked())
 			if err != nil {
 				return err
 			}
-			s.discoverAddressesFinished()
-			s.discoverAccounts = false
```

Comment:
I think we need to restore this `discoverAccounts` property - actually make it a map of walletId-boolean properties. Initial values will be set to `!w.Locked()` but subsequently set to `false` after calling `w.DiscoverActiveAddresses`.